### PR TITLE
Stage Detection 

### PR DESCRIPTION
**Pull request by vardam**
DOES:
- Uses state machines, enums, switch statements, and custom header file to find the current stage of the rocket along with sensor info from the barometer and accelerometer using BMP085 and MPU6050(protoboard).

NEED TO FIX: 
- Probably need to fix a few syntax errors as my VS Code for some reason is not displaying them.
- Need to adjust thresholds based on timings and impulses(after testing)

### CODE REVIEW

File: Gimbalize/src/stage_detection.cpp [hunk 1]
```diff
@@ -1 +1,83 @@
-//Pseudo functions of state machine 
\ No newline at end of file
+#include <Wire.h>
+#include <Adafruit_MPU6050.h>
```
Comment (pchou05) [inline Gimbalize/src/stage_detection.cpp:2]: #include <MKRIMU.h>

File: Gimbalize/src/stage_detection.cpp [hunk 2]
```diff
@@ -1 +1,83 @@
...
+            current_stage = FlightPhase.DESCENT;
+        case (current_stage == FlightPhase.DESCENT && abs(0 - accel_z) < 2) : // <2 is accounting for hallucination data
+            Serial.println("Landing detected!");
+            current_stage = FlightPhase.LANDING;
+        default:
+            Serial.println("PROBLEM DETECTED!");
+        }
+    
+
+    lastPressure = currentPressure;
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.cpp:78]: none of these were declared or initalized

File: Gimbalize/src/stage_detection.cpp [hunk 3]
```diff
@@ -1 +1,83 @@
...
+
+    // Read altitude data from BMP085
+    const altitude = bmp.readAltitude();
+}
+
+// Main loop
+void detectFlightPhase(float accel_x, float accel_y, float accel_z, float altitude, float current_pressure, float last_pressure, float current_stage){
+
+    read_sensor_data(accel_x, accel_y, accel_z, altitude);
+    current_pressure = 1013;
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.cpp:50]: what are these for, why are u changing what we passed?

File: Gimbalize/src/stage_detection.cpp [hunk 4]
```diff
@@ -1 +1,83 @@
...
+    last_pressure = 1013;
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.cpp:51]: same as above

File: Gimbalize/src/stage_detection.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    DESCENT,
+    LANDING
+};
+
+FlightPhase current_stage; 
+
+
+// Function prototypes
+void initializeSensors();
+void readSensorData(float &accel_x, float &accel_y, float &accel_z, float &altitude);
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.h:27]: not using the same function, different names????

File: Gimbalize/src/stage_detection.cpp [hunk 6]
```diff
@@ -1 +1,83 @@
...
+      delay(10);
+    }
+  }
+  Serial.println("BMP085 Found!");
+}
+
+// Function to read sensor data
+void read_sensor_data(float &accel_x, float &accel_y, float &accel_z, float &altitude) {
+    // Read acceleration data from MPU6050
+    mpu.getAcceleration(&accel_x, &accel_y, &accel_z);
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.cpp:40]: declare variable for sensor events

File: Gimbalize/src/stage_detection.cpp [hunk 7]
```diff
@@ -1 +1,83 @@
...
+    Serial.println("Failed to find BMP180 sensor");
+    while (1) {
+      delay(10);
+    }
+  }
+  Serial.println("BMP085 Found!");
+}
+
+// Function to read sensor data
+void read_sensor_data(float &accel_x, float &accel_y, float &accel_z, float &altitude) {
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.cpp:38]: why does this have parameters if your just reading data????

File: Gimbalize/src/stage_detection.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+#include <Wire.h>
+#include <MPU6050.h>
+#include <BMP085.h>
+
+// Define thresholds and other parameters
+const float launch_accel_threshold;
+const float coast_accel_threshold;
+const float apogee_altitude_threshold;
+
+enum FlightPhase {
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.h:13]: use "enum class"

File: Gimbalize/src/stage_detection.h [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    GROUND,
```
Comment (namnik100) [inline Gimbalize/src/stage_detection.h:14]: add comments to explain the significance of each one

**Review by namnik100 — CHANGES_REQUESTED** [review]
Check issues for other things as well

File: Gimbalize/src/stage_detection.cpp (hunk 1 above)
Comment (vardam) [reply Gimbalize/src/stage_detection.cpp:2]: @namnik100 @pchou05 Should I adapt all the code for the mkr imu or keep it for the mpu and bmp?

**Pull request closed without merging**